### DISCUSSION TITLE
Feature Request: Stitches Object Type

### DISCUSSION THREAD

**Discussion by joe-bell** [stitchesjs/stitches#890] [Ideas]
At @monographhq we've recently started exploring the convention of storing our Stitches.js styles in separate `.css.ts` files (similar to [vanilla-extract](https://vanilla-extract.style/))

We do this to:
1.  Get better control of our design system's exports
   _Allowing for styles to be referenced privately/internally_
2. Reference Stitches.js objects inside Storybook
   _e.g. mapping over `variants` and `defaultVariants` to set [options for "controls"](https://storybook.js.org/docs/react/essentials/controls) dynamically_

**The Problem**

Stitches.js offers the `CSS` type for handling basic styles, but as it stands there's no way to obtain the object syntax type (the argument type for `css()` or the second argument type for `styled`) – a combination of the `CSS` type + handling of `variants`, `defaultVariants` and `compoundVariants`

I attempted using TypeScript's built in [`Parameters<typeof styled>`](https://www.typescriptlang.org/docs/handbook/utility-types.html#parameterstype) but had no luck 😢 (it returns `never`) 

**Proposed solution**

It would be awesome if Stitches could provide a helper type (like `CSS` for handling the style object type)

In this example I suggest `StyleObject`, but perhaps `Component` may make more sense?

```ts
// stitches.config.ts
import type * as Stitches from '@stitches/react'
import { createStitches, defaultThemeMap } from '@stitches/react'

const stitchesConfig = createStitches({
  // …config
})

export const { config, styled, css, globalCss, keyframes, getCssText, theme } =
  stitchesConfig

// create type in stitches.config.ts
export type StitchesStyleObject = Stitches.StyleObject<typeof stitchesConfig>
```
```ts
// component.ts
import { styled, StitchesStyleObject } from './stitches.config'

// provides the same inline TypeScript experience as using `styled()`
export const boxConfig: StitchesStyleObject = {
 boxSizing: 'border-box'
}

export const SomeComponent = styled('div', boxConfig) 
```

If this is a little too much to consider right now, I'd really appreciate any guidance on how to achieve this in the meantime 🙏 

---

cc @dknisley @expandrew

**Comment by peduarte** [comment]
Thanks, makes a lot of sense. We'll consider this internally. @hadihallak do you know if there's a current work around for this, even if it means storying some types product-side? Thanks

**Comment by joe-bell** [comment]
Been thinking about this some more over the last few days and I don't think this is actually possible in TypeScript (without defining a function for generics) 😢 

This may be trickier than I anticipated

**Reply by hadihallak** [reply]
yeah having a function generic allows us to extract things like variants that we then pass in deeply to type the JSX calls and that kind of type data would get lost if we use a normal type like in the `StitchesStyleObject` example

**Comment by MrKou47** [comment]
I meet your problem at this time. When I trying to preview Component with variant, like:
```ts
// button.tsx
const Button = styled('button', {
  variants: {
    mode: {
      mode1: {},
      mode2: {}
    }
  }
}
```
```tsx
// button.stories.tsx
import { Button } from './button'

export default {
  component: Button,
}
```
Storybook can't auto generate args, so I should do it manually..

**Reply by andrewdoro** [reply]
@MrKou47 did you find any solution for auto generating args?

**Reply by MrKou47** [reply]
> @MrKou47 did you find any solution for auto generating args?

Nope. I still do this work manually😂

**Reply by andrewdoro** [reply]
@MrKou47  https://github.com/stitchesjs/stitches/issues/239#issuecomment-885293869 I found this approach, but I can't get it work. Do you have any idea?

**Reply by andrewdoro** [reply]
it actually works, the problem I had was that I used numeric values for variants options. Those values are not parsed in the storybook controls, unless I add a letter.
```
 size: {
      "1": {
        borderRadius: "$sm",
        height: "$1",
        px: "$2",
        fontSize: "$sm",
        boxShadow: "$1",
      },
      "2": {
        borderRadius: "$md",
        height: "$5",
        px: "$3",
        fontSize: "$md",
        boxShadow: "$1",
      },
    },
```
